### PR TITLE
Simplified some monster picker logic

### PR DESCRIPTION
**Pull request by FredericF-IT**

### CODE REVIEW

File: src/lib/monsters.lua [hunk 1]
```diff
@@ -2207,7 +2207,6 @@ monster_list = [
         token = "jrcxt",
         size = "M",
         type_str = "Humanoid (human, shapechanger)",
-        size = "M",
```
Comment (Sykander) [inline src/lib/monsters.lua:2210]: nice catch

File: src/lib/encounters.lua [hunk 2]
```diff
@@ -35,36 +40,44 @@ def generate_encounter(target_cr: int, seed: int):
...
+    while cr_budget > 0 and len(final_list) < MONSTER_LIMIT and len(monsters) > 0:
+        cr_list = getRandomFromDict(monsters)
+        mon = cr_list[random.get_random_integer(0, len(cr_list) - 1)]
         cr_budget -= mon.cr
         final_list.append(mon)
 
+        # remove any monster lists with cr above the budget
+        for key in monsters.keys():
+            if key > cr_budget:
+                monsters.pop(key)
```
Comment (Sykander) [inline src/lib/encounters.lua:80]: we should pick which one to remove randomly probably -> fine to do in a later pr

Comment (FredericF-IT) [reply src/lib/encounters.lua:80]: But I mean, if the cr is above the remaining budget, they should never be picked anyways, right?
Or do you mean to remove additional lists that could still be picked?

Comment (Sykander) [reply src/lib/encounters.lua:80]: if we're removing monsters to stay below the cr budget it shouldn't necessarily be the last one that came in i mean

File: src/lib/encounters.lua [hunk 3]
```diff
@@ -35,36 +40,44 @@ def generate_encounter(target_cr: int, seed: int):
...
+        
+    cr_min = max(ceil(cr_min), 0)
 
     monsters = monster_tools.get_monsters_in_cr_range(cr_min, cr_max)
 
-    # if there's no monsters in that cr range search lower
-    while len(monsters) == 0 and cr_min >= 0:
-        cr_min = floor(cr_min - 1)
+    # if there's no monsters in that cr range, expand range
+    while len(monsters) == 0 and cr_min > 0:
```
Comment (Sykander) [inline src/lib/encounters.lua:50]:
```suggestion
    while len(monsters) == 0 and cr_min >= 0:
```

File: src/lib/encounters.lua [hunk 4]
```diff
@@ -35,36 +40,44 @@ def generate_encounter(target_cr: int, seed: int):
...
+        cr_min = cr_min - 1
         monsters = monster_tools.get_monsters_in_cr_range(cr_min, cr_max)
 
-    cr_min = max(cr_min, 0)
```
Comment (Sykander) [inline src/lib/encounters.lua:46]: we should get this line back

File: src/lib/encounters.lua [hunk 5]
```diff
@@ -35,36 +40,44 @@ def generate_encounter(target_cr: int, seed: int):
...
+    # if there is still no monsters, throw exception
+    if len(monsters) == 0:
```
Comment (Sykander) [inline src/lib/encounters.lua:55]: nice addition 🆗

File: src/lib/encounters.lua (hunk 2 above)
Comment (FredericF-IT) [reply src/lib/encounters.lua:80]: I changed the get_monsters_in_cr_range() to return the monsters partitioned into their cr, so you get a dict with the key 1 and the associated entry that contains the list of all cr 1 monsters etc
So this step there would remove the list of monsters of a key (=cr) > cr_budget

Comment (FredericF-IT) [reply src/lib/encounters.lua:80]: I'm still confused about what exactly you mean, but if that is for a future pr, then we can discuss that in a new issue after this pr is reviewed

File: src/lib/encounters.lua (hunk 3 above)
Comment (FredericF-IT) [reply src/lib/encounters.lua:50]: I dont think we want this, as cr_min = 0 would then immediately be reduced to cr_min = cr_min - 1, so -1.

File: src/lib/encounters.lua [hunk 6]
```diff
@@ -35,36 +40,45 @@ def generate_encounter(target_cr: int, seed: int):
     else:
         cr_min = max(target_cr - 3, 0)
         cr_max = target_cr
+        
+    cr_min = max(ceil(cr_min), 0)
```
Comment (Sykander) [inline src/lib/encounters.lua:44]:
```suggestion
    cr_min = max(cr_min, 0)
```

File: src/lib/encounters.lua (hunk 3 above)
Comment (Sykander) [reply src/lib/encounters.lua:50]: if we don't let it drop to cr_min = 0 then we won't let it check for cr 0 monsters - so we need to let cr min go all the way down to our lowest cr monsters

**Review by Sykander — APPROVED** [review]

**Pull request merged**